tests/base/testFunctorHolder.cpp

```diff
@@ -0,0 +1,189 @@
+/**
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU Lesser General Public License as
+ *  published by the Free Software Foundation, either version 3 of the
+ *  License, or  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ **/
+
+/**
+ * @file
+ * @author Roland Denis (\c denis@math.univ-lyon1.fr )
+ * CNRS, Institut Camille Jordan (UMR 5208), Université Lyon 1, France
+ *
+ * @date 2018/07/09
+ *
+ * This file is part of the DGtal library.
+ */
+
+///////////////////////////////////////////////////////////////////////////////
+#include <functional>
+
+#include "DGtal/base/FunctorHolder.h"
+#include "DGtal/base/Common.h"
+#include "DGtal/base/CUnaryFunctor.h"
+
+#include "DGtalCatch.h"
+///////////////////////////////////////////////////////////////////////////////
+
+using namespace DGtal;
+using namespace DGtal::functors;
+using namespace std;
+
+///////////////////////////////////////////////////////////////////////////////
+// Functor and functions
+
+struct Functor
+{
+  double cst;
+
+  explicit Functor(double c) : cst(c) {}
+  inline double operator() (double v) const { return v + cst; }
+};
+
+double add(double v)
+{
+  return v + 1.5;
+}
+
+
+///////////////////////////////////////////////////////////////////////////////
+// Test cases
+
+TEST_CASE( "Holding a lambda", "[lambda]" )
+{
+  double cst = 1.5;
+
+  // Holding an unary lambda by mutable lvalue
+  {
+    auto fn = [&cst] (double v) { return v + cst; };
+    auto holder = holdFunctor(fn);
+    BOOST_CONCEPT_ASSERT(( boost::Assignable<decltype(holder)> ));
+    BOOST_CONCEPT_ASSERT(( concepts::CUnaryFunctor<decltype(holder), double, double> ));
+    REQUIRE( holder(0.5) == 2.0 );
+    auto holder2 = holder;
+    REQUIRE( holder2(0.5) == 2.0 );
+  }
+
+  // Holding an unary lambda by constant lvalue
+  {
+    const auto fn = [&cst] (double v) { return v + cst; };
+    auto holder = holdFunctor(fn);
+    BOOST_CONCEPT_ASSERT(( boost::Assignable<decltype(holder)> ));
+    BOOST_CONCEPT_ASSERT(( concepts::CUnaryFunctor<decltype(holder), double, double> ));
+    REQUIRE( holder(0.5) == 2.0 );
+    auto holder2 = holder;
+    REQUIRE( holder2(0.5) == 2.0 );
```

Comment:
What happened if we return by value fn and assign it to a fn2 for instance ? Is this still working ?

Reply:
You mean, if we return `holder` ?
The problem here is that `holder` refer to `fn` by reference. Each copy of `holder` will refer to the same `fn`. Since `fn` will be destroyed just after the `return`, we obtain a copy of `holder` with a dangling reference to an instance of the lambda (but it may still work since `fn` has no state).
To solve this, we have to move `fn` when calling `holdFunctor` or simply construct it inline:
```auto holder = holdFunctor([&cst] (double v) { return v + cst; });```